packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -2138,6 +2125,118 @@ export class MergeTree {
+     * This primarily affects the ordering of adjacent segments that were removed between the original submission of
+     * the local ops and now.
+     * Consider the following sequence of events:
+     * Initial state: "hi my friend" (seq: 0)
+     * - Client 1 inserts "good " to make "hi my good friend" (op1, refSeq: 0)
+     * - Client 2 deletes "my " to make "hi friend" (op2, refSeq: 0)
+     * - op2 is sequenced giving seq 1
+     * - Client 1 disconnects and reconnects at seq: 1.
+     *
+     * At this point in time, client 1 will have segments ["hi ", Removed"my ", Local"good ", "friend"].
+     * However, the rebased op that it submits will cause client 2 to have segments
+     * ["hi ", Local"good ", Removed"my ", "friend"].
+     *
+     * The difference in ordering can be problematic for tie-breaking concurrently inserted segments in some scenarios.
+     * Rather than incur extra work tie-breaking these scenarios for all clients, when client 1 rebases its operation,
+     * it can fix up its local state to align with what would be expected of the op it resubmits.
+     */
+    public normalizeSegmentsOnRebase(segmentGroup: SegmentGroup): void {
+        // Forward and backward excursions will expand the affected normalization range,
+        // so simply taking the first segment suffices
+        const { refSeq, segments: [segment] } = segmentGroup;
+        const forwardLocalSegs: ISegment[] = [];
+        const backwardLocalSegs: ISegment[] = [];
+        const adjacentRemovedSegs: ISegment[] = [];
+        backwardExcursion(segment, (seg) => {
+            if (seg.seq === UnassignedSequenceNumber) {
+                backwardLocalSegs.push(seg);
+                return true;
+            }
+            const removalInfo = toRemovalInfo(seg);
+            if (removalInfo !== undefined &&
+                removalInfo.removedSeq !== UnassignedSequenceNumber &&
+                removalInfo.removedSeq > refSeq
+            ) {
+                adjacentRemovedSegs.push(seg);
+                return true;
+            }
+            return false;
+        });
+
+        if (adjacentRemovedSegs.length === 0) {
+            return;
+        }
+
+        forwardExcursion(segment, (seg) => {
+            if (seg.seq === UnassignedSequenceNumber) {
+                forwardLocalSegs.push(seg);
+                return true;
+            }
+            return false;
+        });
+        const consecutiveLocalSegments = [...backwardLocalSegs.reverse(), segment, ...forwardLocalSegs];
+        adjacentRemovedSegs.reverse();
+
+        const currentOrder = [...adjacentRemovedSegs, ...consecutiveLocalSegments].map(
+            (seg) => ({
+                parent: seg.parent,
+                index: seg.index,
+                ordinal: seg.ordinal,
+            }),
+        );
+
+        const newOrder = [...consecutiveLocalSegments, ...adjacentRemovedSegs];
+        // TODO: Invoke slide ref callback and test that this happens
```

Comment:
I wanted to merge in Tony's walk references helper for this

Reply:
it's in now